pulp_rpm/app/depsolving.py

```diff
@@ -830,12 +810,246 @@ def run_solver_jobs(jobs):
                 unit_install_job = self._pool.Job(flags, solvable.id)
                 install_jobs.append(unit_install_job)
 
-        # Depsolve using the list of unit install jobs, add them to the results
-        solvables_copied = run_solver_jobs(install_jobs)
-        result_solvables.update(solvables_copied)
+        # Take a list of jobs, get a solution, return the set of solvables that needed to
+        # be installed.
+        solver = self._pool.Solver()
+        solver.set_flag(solv.Solver.SOLVER_FLAG_FOCUS_INSTALLED, 1)
+
+        raw_problems = solver.solve(install_jobs)
+        # The solver is simply ignoring the problems encountered and proceeds associating
+        # any new solvables/units. This might be reported back to the user one day over
+        # the REST API. For now, log only "real" dependency issues (typically some variant
+        # of "can't find the package"
+        dependency_warnings = self._build_warnings(raw_problems)
+        if dependency_warnings:
+            logger.warning(
+                "Encountered problems solving dependencies, "
+                "copy may be incomplete: {}".format(", ".join(dependency_warnings))
+            )
+
+        transaction = solver.transaction()
+        if settings.SOLVER_DEBUG_LOGS:
+            write_solver_debug_data(solver, raw_problems, self.mapping, full=False)
+        result_solvables.update(set(transaction.newsolvables()))
 
         solved_units = self.mapping.get_units_from_solvables(result_solvables)
         for k in unit_repo_map.keys():
             solved_units[k] |= passthrough[k]
 
         return solved_units
+
+
+# Descriptions copied from the libsolv documentation
+# https://github.com/openSUSE/libsolv/blob/master/doc/libsolv-bindings.txt
+def write_solver_debug_data(solver, problems, mapping, full=False):
+    """Dump the state of the solver including actions decided upon and problems encountered."""
+    from pulpcore.plugin.models import Task
+    from pathlib import Path
+
+    debugdata_dir = Path("/var/tmp/pulp") / str(Task.current().pulp_id)
+    debugdata_dir.mkdir(parents=True, exist_ok=True)
+    logger.info("Writing solver debug data to {}".format(debugdata_dir))
+
+    transaction = solver.transaction()
+    summary_path = debugdata_dir / "depsolving_summary.txt"
+
+    def reason_desc(reason):
```

Comment:
I think reason_desc() and rule_desc() could be replaced with {key: (tuple)} dictionaries. If those could be external files somehow, would make it easier to fiddle with text without having to change code. Also, do we want to consider I18N for the descriptions?

Reply:
re: dictionaries, good idea.

re: i18n, maybe.  This is only for debugging purposes so I'm not sure if it would be of much benefit?  Also how are translations actually done?

Reply:
We don't have L10N processes in place yet. Having the description-strs be gettext-ed (ie, surrounded by _() ) would make it possible to localize at some point.

As you say tho, this is for advanced-debugging, and anyone who'd be looking at it is already used to english-descriptions from libsolv code anyway.